Add Time-LLM

marcopeix opened the pull request:
Add TimeLLM to neuralforecast.

User can load an LLM from HugginFace and use it for time series forecasting.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Nixtla/neuralforecast?pullRequest=908) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Nixtla/neuralforecast?pullRequest=908) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: marcopeix<br/>:x: Marco Peixeiro<br/><hr/>**Marco Peixeiro** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Nixtla/neuralforecast?pullRequest=908) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Nixtla/neuralforecast?pullRequest=908) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Nixtla/neuralforecast?pullRequest=908) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: marcopeix<br/>:x: Marco Peixeiro<br/><hr/>**Marco Peixeiro** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Nixtla/neuralforecast?pullRequest=908) it.</sub>

cchallu commented on nbs/models.timellm.ipynb:
```diff
@@ -0,0 +1,940 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #74.                     llm = None,
</pre><p> </p><p>If it is needed, remove default of None. Also, it is not clear from the docstring how to use it, what do users need to pass?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Nixtla/neuralforecast/pull/908/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

marcopeix replied:
Done! I added GPT-2 as the default LLM if the user does not specify another model.

marcopeix commented:
I added tests to the notebook
GPT-2 is the default LLM is the user does not specify another model
The usage example acts as a tutorial for using Time-LLM
Added `transformers` as a dependency

AzulGarza commented on experiments/test_timellm/test_timellm.py:
```diff
@@ -0,0 +1,55 @@
+import os 
```
this test is also part of `models.timellm.ipynb`, could we remove this file to keep the unit tests in the nbs?

AzulGarza commented on neuralforecast/models/timellm.py:
```diff
@@ -0,0 +1,497 @@
+        llm_num_hidden_layers=32,
+        llm_output_attention: bool = True,
+        llm_output_hidden_states: bool = True,
+        prompt_prefix: str = None,
```
adding `propmpt_prefix: Optional[str] = None` should fix the [linter](https://github.com/Nixtla/neuralforecast/actions/runs/8142491133/job/22252164542?pr=908) errors.

AzulGarza reviewed (changes requested):
thanks @marcopeix! 

could you also include `transformers` in the `dev_requirements` part of the [`setttings.ini`](https://github.com/Nixtla/neuralforecast/blob/10bb2e2f861d511123ec95646c655f2403da5f92/settings.ini#L19) file? 

also running `nbdev_install_git_hooks ` and `nbdev_clean` should solve the [`nb_sync` action](https://github.com/Nixtla/neuralforecast/actions/runs/8142491138/job/22252165334?pr=908).

AzulGarza commented on neuralforecast/models/timellm.py:
```diff
@@ -0,0 +1,497 @@
+
+from ..losses.pytorch import MAE
+
+from transformers import GPT2Config, GPT2Model, GPT2Tokenizer
```
since `transfomers` is not a core dependency, we can try to see if it is installed and then throw an error when it needs to be used for default arguments (`llm=None`), something like:

```python
IS_TRANSFORMERS_INSTALLED = True
try:
     from transformers import GPT2Config, GPT2Model, GPT2Tokenizer
except ImportError:
     IS_TRANSFORMERS_INSTALLED = False

AzulGarza commented on neuralforecast/models/timellm.py:
```diff
@@ -0,0 +1,497 @@
+    `n_heads`: int=8, number of heads in attention layer.<br>
+    `enc_in`: int=7, encoder input size.<br>
+    `dec_in`: int=7, decoder input size.<br>
+    `llm` = None, LLM model to use. For example, you can use GPT-2 from https://huggingface.co/openai-community/gpt2"<br>
```
could we document the default behavior when `llm=None`?

AzulGarza commented on neuralforecast/models/timellm.py:
```diff
@@ -0,0 +1,497 @@
+        self.llm = llm
+        self.llm_tokenizer = llm_tokenizer
+
+        if self.llm_config is None:
+            self.llm_confg = GPT2Config.from_pretrained("openai-community/gpt2")
+        if self.llm is None:
+            self.llm = GPT2Model.from_pretrained(
+                "openai-community/gpt2", config=self.llm_confg
+            )
+        if self.llm_tokenizer is None:
+            self.llm_tokenizer = GPT2Tokenizer.from_pretrained("openai-community/gpt2")
```
i think it might be better if we use `GPT2` when `llm is None`, I'm thinking for example, what happens if `llm_tokenizer is not None` but for some reason `llm is None`? also, we should add an exception based on `IS_TRANSFORMERS_INSTALLED` in case that the default params are being used, something like:

```python
if self.llm is None:
      if not IS_TRANSFORMERS_INSTALLED:
             raise ImportError("please install `transformers` to use default llm")
      print("Using GPT2 model as default and ignoring `llm_config` and `llm_tokenizer`")
      self.llm = GPT2...
      self.llm_config = GPT2...
      self.llm_tokenizer = GPT2...
```

cchallu commented:
I don't think llm = None is a good idea. Either put the default model or nothing, like h or input size so users know they have to fill it.

marcopeix commented:
Updated PR according to Azul's comments:
- Better documentation of default behaviour
- Remove transformers from core dependencies and added to dev dependencies
- Flag to check if `transformers` is installed and use GPT-2 as default LLM
- Remove test file from experiments folder since the script exists in the notebook

AzulGarza reviewed (approved):
awesome @marcopeix! thank you 🎉

let's wait for the review of @jmoralez or @cchallu to merge it to main❤️
